### PR TITLE
Winner tab is not selected when user click on +X users from 'Past cha…

### PR DESCRIPTION
**Pull request by daft300punk**
…llenges' #206 

1. Tab selection is now handled by global redux state.
2. For SSR, we set the state for selected tab in the related reducer factory.

To Test Locally - 
1. Goto http://localhost:3000/challenges/30050696 will open the page with details tab, i.e details tab is default selected.
2. Goto http://localhost:3000/challenges/30050696?tab=registrants will open the page with registrants tab selected.
3. Note http://localhost:3000/challenges/30050696?tab=checkpoints causes a problem at the moment, since the Checkpoint related component uses window. As part of another issue #170 - I've removed dependency on window, so it should get solved with that merge.
4. After the page is loaded for the first time, subsequent tab selection is handled on the client via redux.

### CODE REVIEW

**Comment by birdofpreyru** [comment]
@daft300punk Any update?

**Comment by daft300punk** [comment]
I'll be updating in a few hours. Was a little away for visa purposes.

File: src/shared/components/challenge-listing/ChallengeCard/Status/index.jsx [hunk 1]
```diff
@@ -379,7 +379,7 @@ class ChallengeStatus extends Component {
         </div>);
     });
     return leaderboard || (
-      <a href={`${challengeURL}${id}#winner`}>Results</a>
+      <a href={`${challengeURL}${id}?tab=winner`}>Results</a>
```
[the remark this answers is not in the document]
Comment (daft300punk) [reply src/shared/components/challenge-listing/ChallengeCard/Status/index.jsx]: I shouldn't have made this change. This is unrelated. challengeURL and id are used here. The link is to old topcoder page.
But as suggested, I now use Link in the related LeaderboardAvatar component, so now app won't be reloaded.

File: src/shared/reducers/challenge.js [hunk 2]
```diff
@@ -217,7 +241,8 @@ export function factory(req) {
    * should be re-used there. */
   /* TODO: For completely server-side rendering it is also necessary to load
    * terms, results, etc. */
-  if (req && req.url.match(/^\/challenges\/\d+$/)) {
+  const urlWithoutQuery = req && req.url.split('?')[0];
+  if (req && urlWithoutQuery.match(/^\/challenges\/\d+$/)) {
```
[the remark this answers is not in the document]
Comment (daft300punk) [reply src/shared/reducers/challenge.js]: Upadted as suggested. Express does include query string in the url.

**Comment by birdofpreyru** [comment]
@daft300punk I've pushed some updates into your PR, but it still does not work correctly and needs further modifications. The main problem is that when you start travelling between pages using `<Link>` components, your code does not update Redux state, and just mentioning `tab=...` in the link does not work in that case due to number of reasons. I.e. if you go to a challenge details page, select, say "registrants" tab, then go back to the challenge listing without reloading the app, and click on +1, which should get you to the winners tab of challenge details, it will get you to "registrants" instead.

To make it work, we need to update your code to be similar with filters in challenge listings (they are also deep-linkable, and serializable into URL). Let me briefly walk you through related code:

- https://github.com/topcoder-platform/community-app/blob/develop/src/shared/reducers/challenge-listing/index.js#L169
  This is reducer that switches challenge listing filters to the specified state. Note that it calls `updateQuery(..)` to serialize filter state into URL query each time when that state is updated (all such updates trigger this reducer).

- https://github.com/topcoder-platform/community-app/blob/develop/src/shared/reducers/challenge-listing/index.js#L284
  Here, during server-side rendering, we check filters state written into query params, and set proper redux state, reusing the same handler referenced above.

- https://github.com/topcoder-platform/community-app/blob/develop/src/shared/components/challenge-detail/Header/ChallengeTags.jsx#L58
  And here is an example, how to switch to the challenge listing with specified filters on the frontend side, without reloading the app (this piece of code relates to technology tags in the header of challenge details page - they bring us to the challenge listing, with filtering by that tech activated). `<Tag>` under the hood, is rendered with `<Link>` from react-router. `to="/challenges"` specifies the route where we want to move (in your case it will be `to="/challenge/12345"`), and the setting of filters (in your case - selection of the tab) is done via `onClick` prop. There `setChallengeListingFilter(..)` dispatches `setFilter` action, which triggers reducer referenced above, which (1) updates query params in URL; (2) updates redux state to set proper filter. The call is wrapped inside `setImmediate(..)` to postpone its execution just after `react-router` handles the transition (because `react-router` will reset query params).

So, doing it this way is good, because the actual logic behind tab switching and query params update is handled all in one place - corresponding reducer, and it is a way less error prone.

### Regarding the Bug Bash

I am targeting to finalize it and pay the earning as soon as possible. If you don't make these fixes (and I don't hear any feedback from you) before I finish reviewing / cleaning up other pending PRs, then I don't count this issue within Bug Bash.

If you keep on working and finish this task, as I have explained above, I'll make you the task payment separately afterwards. Alternatively, I can finish it myself, paying you a reduced prize for your efforts here (I'd say $80). Let me know what you think.

**Comment by daft300punk** [comment]
@birdofpreyru I won't be able to finish this. I'm fine with reduced payment. Thanks.

**Pull request merged**